src/contract-abi.ts

```diff
@@ -0,0 +1,340 @@
+  name: string;
+}
+
+export interface ClarityAbiTypeNonFungibleToken {
+  name: string;
+  type: ClarityAbiType;
+}
+
+export interface ClarityAbi {
+  functions: ClarityAbiFunction[];
+  variables: ClarityAbiVariable[];
+  maps: ClarityAbiMap[];
+  fungible_tokens: ClarityAbiTypeFungibleToken[];
+  non_fungible_tokens: ClarityAbiTypeNonFungibleToken[];
+}
+
+function matchType(cv: ClarityValue, abiType: ClarityAbiType): boolean {
+  const union = getTypeUnion(abiType);
+
+  switch (cv.type) {
+    case ClarityType.BoolTrue:
+    case ClarityType.BoolFalse:
+      return union.id === ClarityAbiTypeId.ClarityAbiTypeBool;
+    case ClarityType.Int:
+      return union.id === ClarityAbiTypeId.ClarityAbiTypeInt128;
+    case ClarityType.UInt:
+      return union.id === ClarityAbiTypeId.ClarityAbiTypeUInt128;
+    case ClarityType.Buffer:
+      return (
+        union.id === ClarityAbiTypeId.ClarityAbiTypeBuffer &&
+        union.type.buffer.length === cv.buffer.length
+      );
+    case ClarityType.OptionalNone:
+      return (
+        union.id === ClarityAbiTypeId.ClarityAbiTypeNone ||
+        union.id === ClarityAbiTypeId.ClarityAbiTypeOptional
+      );
+    case ClarityType.OptionalSome:
+      return (
+        union.id === ClarityAbiTypeId.ClarityAbiTypeOptional &&
+        matchType(cv.value, union.type.optional)
+      );
+    case ClarityType.ResponseErr:
+      return (
+        union.id === ClarityAbiTypeId.ClarityAbiTypeResponse &&
+        matchType(cv.value, union.type.response.error)
+      );
+    case ClarityType.ResponseOk:
+      return (
+        union.id === ClarityAbiTypeId.ClarityAbiTypeResponse &&
+        matchType(cv.value, union.type.response.ok)
+      );
+    case ClarityType.PrincipalContract:
+    case ClarityType.PrincipalStandard:
+      return union.id === ClarityAbiTypeId.ClarityAbiTypePrincipal;
+    case ClarityType.List:
+      return (
+        union.id == ClarityAbiTypeId.ClarityAbiTypeList &&
+        union.type.list.length === cv.list.length &&
+        _.every(cv.list, val => matchType(val, union.type.list.type))
+      );
+    case ClarityType.Tuple:
+      if (union.id == ClarityAbiTypeId.ClarityAbiTypeTuple) {
+        const tuple = _.cloneDeep(cv.data);
+        for (let i = 0; i < union.type.tuple.length; i++) {
+          const abiTupleEntry = union.type.tuple[i];
+          const key = abiTupleEntry.name;
+          const val = tuple[key];
+
+          // if key exists in cv tuple, check if its type matches the abi
+          // return false if key doesn't exist
+          if (val) {
+            if (!matchType(val, abiTupleEntry.type)) {
+              return false;
+            }
+            delete tuple[key];
+          } else {
+            return false;
+          }
+        }
+        return true;
+      }
+    default:
+      return false;
+  }
+}
+
+export function validateContractCall(payload: ContractCallPayload, abi: ClarityAbi): boolean {
```

Comment:
Would also help to have some documentation if we expect developers to use this.

Reply:
Adding documentation, however, I don't think we expect developers to use either of these functions. these functions get called in the `makeContractCall` builder function based on the options that are provided. For example, if you pass in a `network` and set `validateWithAbi: true`, then it will attempt to use `getAbi` to fetch the ABI and validate with it. Otherwise, you can pass in a `ClarityAbi` object to `validateWithAbi` and it will skip the `getAbi` request and use the provided ABI. I'll add this info to the relevant section of the docs.